UserData wrappers for CloudFormation helper scripts when using Beanstalk

jawadst opened the pull request:
# Issue Type

- Feature Pull Request

## Summary

Fixes #344 

This PR implements the fix suggested by AWS support: force the CloudFormation helper scripts (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-helper-scripts-reference.html) to use the instance role instead of the standard CloudFormation authentication protocol.

This is implemented in a few steps:

1. Global configuration flag (or per AS group) to enable the feature: `beanstalk_cfn_init_role`
2. When enabled and when AutoSpotting launches a new instance: check if the `UserData` of the instance belongs to Beanstalk and, if it does, alter the `UserData` to add wrappers (bash scripts)

The wrappers are done as follow:
- `cfn-init`, `cfn-get-metadata` and `cfn-signal` are wrapped by adding the instance role to the original code as `--role instance-role`
- `cfn-hup` does not accept a `--role` param so we write the role into the config file /etc/cfn/cfn-hup.conf

## Confirm that the fix works

For record keeping (for future debugging or if anyone has the same issue), here is what I am looking at to confirm that everything works:

1) The spot instance launched by AutoSpotting should have its UserData patched.

**Example of unpacted user data:**
```bash
Content-Type: multipart/mixed; boundary="===============5189065377222898407=="
MIME-Version: 1.0

--===============5189065377222898407==
Content-Type: text/cloud-config; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="cloud-config.txt"

#cloud-config
repo_upgrade: none
repo_releasever: 2018.03
cloud_final_modules:
 - [scripts-user, always]

--===============5189065377222898407==
Content-Type: text/x-shellscript; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="user-data.txt"

#!/bin/bash
exec > >(tee -a /var/log/eb-cfn-init.log|logger -t [eb-cfn-init] -s 2>/dev/console) 2>&1
echo [`date -u +"%Y-%m-%dT%H:%M:%SZ"`] Started EB User Data
set -x

(...)
```

**Example of patched user data:**
```bash
Content-Type: multipart/mixed; boundary="===============5189065377222898407=="
MIME-Version: 1.0

--===============5189065377222898407==
Content-Type: text/cloud-config; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="cloud-config.txt"

#cloud-config
repo_upgrade: none
repo_releasever: 2018.03
cloud_final_modules:
 - [scripts-user, always]

--===============5189065377222898407==
Content-Type: text/x-shellscript; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename="user-data.txt"

#!/bin/bash
---- modify CloudFormation helpers ----
# Modify cfn-init to use --role by default
echo -e '#!/bin/bash\nINROLE=$(curl -s 169.254.169.254/latest/meta-data/iam/security-credentials/)\n/opt/aws/bin/cfn-init-2 --role $INROLE "$@" \nexit $?' > /opt/aws/bin/cfn-init.tmp
mv /opt/aws/bin/cfn-init /opt/aws/bin/cfn-init-2
mv /opt/aws/bin/cfn-init.tmp /opt/aws/bin/cfn-init
chmod +x /opt/aws/bin/cfn-init

# Modify cfn-get-metadata to use --role by default
echo -e '#!/bin/bash\nINROLE=$(curl -s 169.254.169.254/latest/meta-data/iam/security-credentials/)\n/opt/aws/bin/cfn-get-metadata-2 --role $INROLE "$@" \nexit $?' > /opt/aws/bin/cfn-get-metadata.tmp
mv /opt/aws/bin/cfn-get-metadata /opt/aws/bin/cfn-get-metadata-2
mv /opt/aws/bin/cfn-get-metadata.tmp /opt/aws/bin/cfn-get-metadata
chmod +x /opt/aws/bin/cfn-get-metadata

# Modify cfn-signal to use --role by default
echo -e '#!/bin/bash\nINROLE=$(curl -s 169.254.169.254/latest/meta-data/iam/security-credentials/)\n/opt/aws/bin/cfn-signal-2 --role $INROLE "$@" \nexit $?' > /opt/aws/bin/cfn-signal.tmp
mv /opt/aws/bin/cfn-signal /opt/aws/bin/cfn-signal-2
mv /opt/aws/bin/cfn-signal.tmp /opt/aws/bin/cfn-signal
chmod +x /opt/aws/bin/cfn-signal

# Modify cfn-hup to use --role by default
echo -e '#!/bin/bash\nprintf "role=$(curl -s 169.254.169.254/latest/meta-data/iam/security-credentials/)" >> /etc/cfn/cfn-hup.conf\n/opt/aws/bin/cfn-hup-2 "$@" \nexit $?' > /opt/aws/bin/cfn-hup.tmp
mv /opt/aws/bin/cfn-hup /opt/aws/bin/cfn-hup-2
mv /opt/aws/bin/cfn-hup.tmp /opt/aws/bin/cfn-hup
chmod +x /opt/aws/bin/cfn-hup
---- modify CloudFormation helpers ----

exec > >(tee -a /var/log/eb-cfn-init.log|logger -t [eb-cfn-init] -s 2>/dev/console) 2>&1
echo [`date -u +"%Y-%m-%dT%H:%M:%SZ"`] Started EB User Data
set -x

(...)
```

2) The initialization process of the Beanstalk instance should complete successfully and without any delay

**Example of system log from incorrect Beanstalk initialization:**

```
[eb-cfn-init]: [2019-10-11T14:22:59.878Z] First init of instance.
[eb-cfn-init]: [2019-10-11T14:22:59.881Z] Started executing cfn_init _OnInstanceBoot first_init.
[eb-cfn-init]: [2019-10-11T14:22:59.885Z] Running cfn-init ConfigSet: _OnInstanceBoot.
[eb-cfn-init]: 11 Oct 14:23:07 ntpdate[2103]: step time server 131.188.3.220 offset 0.558345 sec
[eb-cfn-init]: Starting ntpd: [  OK  ]

[eb-cfn-init]: [2019-10-11T14:27:22.748Z] Command Returned: 
[eb-cfn-init]: [2019-10-11T14:27:22.750Z] Command return code 1.
[eb-cfn-init]: [2019-10-11T14:27:22.753Z] Tailing /var/log/eb-commandprocessor.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-commandprocessor taillog *******************
[eb-cfn-init]: ***eb-commandprocessor is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.760Z] Tailing /var/log/eb-activity.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-activity taillog *******************
[eb-cfn-init]: ***eb-activity is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.767Z] Tailing /var/log/eb-tools.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-tools taillog *******************
[eb-cfn-init]: ***eb-tools is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.773Z] Tailing /var/log/eb-version-deployment.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-version-deployment taillog *******************
[eb-cfn-init]: ***eb-version-deployment is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.775Z] Tailing /var/log/cfn-init.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-init taillog *******************
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:25:56,291 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:16,158 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:27,649 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:36,765 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:55,459 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:13,114 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:19,372 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.836Z] Tailing /var/log/cfn-hup.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-hup taillog *******************
[eb-cfn-init]: ***cfn-hup is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:22.842Z] Sleeping 10 ...
[eb-cfn-init]: [2019-10-11T14:27:32.846Z] Retrying...
[eb-cfn-init]: [2019-10-11T14:27:32.848Z] Running cfn-init ConfigSet: _OnInstanceBoot.
[eb-cfn-init]: [2019-10-11T14:27:33.196Z] Command Returned: 
[eb-cfn-init]: [2019-10-11T14:27:33.199Z] Command return code 1.
[eb-cfn-init]: [2019-10-11T14:27:33.201Z] Tailing /var/log/eb-commandprocessor.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-commandprocessor taillog *******************
[eb-cfn-init]: ***eb-commandprocessor is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.208Z] Tailing /var/log/eb-activity.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-activity taillog *******************
[eb-cfn-init]: ***eb-activity is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.214Z] Tailing /var/log/eb-tools.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-tools taillog *******************
[eb-cfn-init]: ***eb-tools is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.218Z] Tailing /var/log/eb-version-deployment.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-version-deployment taillog *******************
[eb-cfn-init]: ***eb-version-deployment is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.222Z] Tailing /var/log/cfn-init.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-init taillog *******************
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:25:56,291 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:16,158 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:27,649 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:36,765 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:55,459 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:13,114 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:19,372 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.282Z] Tailing /var/log/cfn-hup.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-hup taillog *******************
[eb-cfn-init]: ***cfn-hup is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:33.288Z] Sleeping 20 ...
[eb-cfn-init]: [2019-10-11T14:27:53.292Z] Retrying...
[eb-cfn-init]: [2019-10-11T14:27:53.294Z] Running cfn-init ConfigSet: _OnInstanceBoot.
[eb-cfn-init]: [2019-10-11T14:27:53.640Z] Command Returned: 
[eb-cfn-init]: [2019-10-11T14:27:53.643Z] Command return code 1.
[eb-cfn-init]: [2019-10-11T14:27:53.645Z] Tailing /var/log/eb-commandprocessor.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-commandprocessor taillog *******************
[eb-cfn-init]: ***eb-commandprocessor is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.652Z] Tailing /var/log/eb-activity.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-activity taillog *******************
[eb-cfn-init]: ***eb-activity is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.658Z] Tailing /var/log/eb-tools.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-tools taillog *******************
[eb-cfn-init]: ***eb-tools is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.665Z] Tailing /var/log/eb-version-deployment.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* eb-version-deployment taillog *******************
[eb-cfn-init]: ***eb-version-deployment is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.671Z] Tailing /var/log/cfn-init.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-init taillog *******************
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:25:56,291 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:16,158 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:27,649 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:36,765 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:26:55,459 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:13,114 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: 2019-10-11 14:27:19,372 [ERROR] Throttling: Rate exceeded
[eb-cfn-init]: Traceback (most recent call last):
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 162, in _retry
[eb-cfn-init]:     return f(*args, **kwargs)
[eb-cfn-init]:   File "/usr/lib/python2.7/dist-packages/cfnbootstrap/util.py", line 234, in _timeout
[eb-cfn-init]:     raise exc[0]
[eb-cfn-init]: HTTPError: 400 Client Error: Bad Request
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.726Z] Tailing /var/log/cfn-hup.log
[eb-cfn-init]: 
[eb-cfn-init]: ******************* cfn-hup taillog *******************
[eb-cfn-init]: ***cfn-hup is not available yet.***
[eb-cfn-init]: ******************* End of taillog *******************
[eb-cfn-init]: 
[eb-cfn-init]: 
[eb-cfn-init]: [2019-10-11T14:27:53.732Z] Sleeping 40 ..
````

Notice the multiple 400/Bad Request errors, and the Beanstalk script sleeping at the end.  
Without patching the user data, after the instance joins the ASG, it becomes authorized and the initialization process finishes. This sometimes fails or is very slow as the retry logic seems to use exponential backoff.

**Example of system log from correct Beanstalk initialization:**

I can't provide a full log as it has a lot of account info but it basically does not have the same errors and the initialization happens without delay.

There is a single 403 error but that I also saw in the system log of regular on-demand instances launched by Beanstalk so I am assuming they are normal/expected:
```
[eb-cfn-init]: Error signaling CloudFormation: [Errno 403] HTTP Error 403 : <?xml version="1.0" encoding="UTF-8"?>
[eb-cfn-init]: <Error><Code>AccessDenied</Code><Message>Request has expired</Message><X-Amz-Expires>86399</X-Amz-Expires><Expires>2017-08-31T18:20:03Z</Expires><ServerTime>2019-10-11T17:47:11Z</ServerTime><RequestId>4E9D6B17B4311FBA</RequestId><HostId>L/GuQ0u1MCagiOfAPrfPmuPqnRB7kgG+s47dYTWiIyU8bJ9F0lKkRVaTwNiatl3oFUP+aHqbh+o=</HostId></Error>
```

## Code contribution checklist

1. [x] The contribution fixes a single existing github issue, and it is linked
   to it.
1. [x] The code is as simple as possible, readable and follows the idiomatic Go
  [guidelines](https://golang.org/doc/effective_go.html).
1. [x] All new functionality is covered by automated test cases so the overall
  test coverage doesn't decrease.
1. [x] No issues are reported when running `make full-test`.
1. [x] Functionality not applicable to all users should be configurable.
1. [x] Configurations should be exposed through Lambda function environment
   variables which are also passed as parameters to the
   [CloudFormation](https://github.com/cristim/autospotting/blob/master/cloudformation/stacks/AutoSpotting/template.yaml)
   and
   [Terraform](https://github.com/autospotting/terraform-aws-autospotting/main.tf)
   stacks defined as infrastructure code.
1. [x] Global configurations set from the infrastructure stack level should also
   support per-group overrides using tags.
1. [x] Tags names and expected values should be similar to the other existing
   configurations.
1. [x] Both global and tag-based configuration mechanisms should be tested and
  proven to work using log output from various test runs.
1. [x] The logs should be kept as clean as possible (use log levels as
  appropriate) and formatted consistently to the existing log output.
1. [x] The documentation is updated to cover the new behavior, as well as the
   new configuration options for both stack parameters and tag overrides.
1. [ ] A code reviewer reproduced the problem and can confirm the code
  contribution actually resolves it.

jawadst commented:
@cristim 
Here is the PR for #344 

I still need to add documentation and expose the new configuration option (`beanstalk_cfn_init_role`) to Lambda/Terraform. The code and tests are ready though so it'd be great if you can take a first look.

I did run this branch against a Beanstalk app on AWS and validated that the fix works as expected by looking at the following items:

- `UserData` of instances launched by AutoSpotting are correctly patched
- No 400 errors in the logs and the initialization of the instance wrt Beanstalk is successful ; there is one error remaining related to `cfn-hup` when it gets launched but I see the same error with on-demand instances launched directly by Beanstalk/ASG so I think it's normal/acceptable
- The instance joins the Beanstalk app before it gets added to the ASG (that was the main issue as it had to wait until it joined the ASG to be able to finish its initialization so 10+ minutes)

jawadst commented:
@cristim 
I have updated the documentation and added configuration in the Lambda template and the Kubernetes example.

Two questions for you:
1) One point I don't understand is how env variables get passed to the`run` function for both Lambda and Kubernetes.

Is it that `flag.StringVar` gets the value from the env variable or the command-line automatically?
```
flag.StringVar(&c.BeanstalkCFNInitRole, "beanstalk_cfn_init_role", "", "\n\tControls whether AutoSpotting patches Elastic Beanstalk UserData scripts to use the instance role when calling CloudFormation helpers instead of the standard CloudFormation authentication method\n"+
		"\tExample: ./AutoSpotting --beanstalk_cfn_init_role true\n")
```

I defined the env variable `BEANSTALK_CFN_INIT_ROLE`.

2) Is there an easy way to test the Lambda template to ensure that it works as expected?  
Is it as simple as doing a custom build as documented on https://github.com/AutoSpotting/AutoSpotting/blob/master/CUSTOM_BUILDS.md?

cristim commented:
Thanks for your contribution, I will have a look at it by the end of the week.

To answer your questions:

1. That's exactly how it works, the flag library we use is automatically handling the  `FOO=bar` environment variable as if using `-foo=bar` in the command line arguments.

2. Yes, just that you will also need to patch the Cloudformation template to pass this environment variable.

I also have a question for you: does the code assume this IAM role exists, being created in advance by the user? I'm thinking that we can perhaps ship it out of the box as part of the Cloudformation template and wire it automatically to the Lambda function environment variables. The only thing we would handle differently would be to patch the userdata script like you are doing, to make use of it. Everyone else would have this IAM role exist but unused.

What do you think about this?

jawadst commented:
@cristim 
The role used is the one from the instance, collected from the instance metadata. Beanstalk requires instances to have one (https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/concepts-roles-instance.html) so it is safe to assume that there is one. Beanstalk creates a default role `aws-elasticbeanstalk-ec2-role` for all environments the first time it gets used in an account. The role being used by the instance could be either that default role or a custom role created by the user.

We could pre-create two things in the CloudFormation template:
- A policy (`AutoSpottingElasticBeanstalk`) that contains the permissions listed in the documentation. The documentation would tell users to add that policy to the `aws-elasticbeanstalk-ec2-role` role manually.
- An IAM role (`autospotting-elasticbeanstalk-ec2-role`) that can be used as a replacement of the default Beanstalk role and that includes the `AutoSpottingElasticBeanstalk` policy as well as other required policies by Beanstalk. The documentation could tell users to use this role when creating new Beanstalk envs / replace the role of their current Beanstalk env.

I am not sure both are required and I think #1 (the policy) is enough and less intrusive.

We won't be able to add the `AutoSpottingElasticBeanstalk` policy to the `aws-elasticbeanstalk-ec2-role` role automatically through CloudFormation.  
The AutoSpotting binary could include some logic at runtime to check if the default role has that policy and, if not, add it. We would need to do an IAM API request to check if the role already exists and if we can add the policy to it.  
Not sure if it's a good idea to automatically modify that role for the user though, as you will now require the AutoSpotting Lambda to have permissions for modifying IAM roles (but we could keep it specific to that IAM role I guess). I am also not sure if AutoSpotting users will like that we modify some of their roles for them (it's convenient but intrusive).

Let me know your thoughts.

cristim commented:
Thanks for the explanation, I am in favor of the policy that people can add to the configuration without much effort

jawadst commented:
@cristim 
I have added a AutoSpottingElasticBeanstalk managed policy. I have also renamed the config parameter from `beanstalk_cfn_init_role` to `beanstalk_cfn_wrappers` to make it clearer and it now accepts an On/Off value (default to Off).

I ran the template in our account from our custom build and tested the following:
- When disabled (`beanstalk_cfn_wrappers = off`), spot instances are launched without any modification to their UserData
- When enabled (`beanstalk_cfn_wrappers = on`) and with the policy added to the role of the instances, UserData are updated and the spot instances launch correctly and join the Beanstalk cluster as soon as they are up

The behavior looks good to me overall.

cristim commented on autospotting.go:
```diff
@@ -215,6 +217,8 @@ func (c *cfgData) parseCommandLineFlags() {
 		"Allowed values: evaluation|I_am_supporting_it_on_Patreon|I_contributed_to_development_within_the_last_year|I_built_it_from_source_code\n"+
 		"\tExample: ./AutoSpotting --license evaluation\n")
+	flag.StringVar(&c.BeanstalkCFNWrappers, "beanstalk_cfn_wrappers", "", "\n\tControls whether AutoSpotting patches Elastic Beanstalk UserData scripts to use the instance role when calling CloudFormation helpers instead of the standard CloudFormation authentication method\n"+
+		"\tExample: ./AutoSpotting --beanstalk_cfn_wrappers true\n")
```
this needs to ne updated to `On`, unless we decide to keep the boolean values as they are as per my other comment

cristim commented on START.md:
```diff
@@ -258,6 +301,13 @@ Usage of ./AutoSpotting:
   -tag_filters=[{spot-enabled true}]: Set of tags to filter the ASGs on.  Default is -tag_filters 'spot-enabled=true'
         Example: ./AutoSpotting -tag_filters 'spot-enabled=true,Environment=dev,Team=vision'
+
+  -beanstalk_cfn_wrappers="on":
```
I would propose to rename this to something like `patch_beanstalk_userdata`, and then we can also keep the true/false boolean values

cristim commented on core/beanstalk_userdata_example.txt:
```diff
@@ -0,0 +1,54 @@
+Content-Type: multipart/mixed; boundary="===============5189065377222898407=="
```
I wouldn't put these files here, I propose to create a test_data directory in the root dir of the repo and move them there

cristim commented:
Looks great, I just had a couple of small comments, but other than that we're pretty close.

Thank you for this high quality contribution, I really appreciate it!

jawadst commented:
@cristim 
Sure, happy to contribute here. We had to disable AutoSpotting because of that CloudFormation update and we are pretty eager to be able to turn it back on so it's only natural that we help.

I have pushed commits that implement your comments. I have run the latest version through CloudFormation and it works correctly.

I have also updated the PR description with how to check whether the patching works and Beanstalk correctly initializes so that we have a clear way of debugging if ever needed.

hash167 commented on core/beanstalk.go:
```diff
@@ -0,0 +1,80 @@
+	decodedUserData := decodeUserData(userData)
+
+	// Patch the UserData if possible
+	if strings.Contains(*decodedUserData, "ebbootstrap") {
```
Instead of just looking for beanstalk data, if we can also look for a tag on an ASG, then maybe regular autoscaling group instance user data will also be patched?

cristim replied:
I don't think we can patch arbitrary userdata, we can do it for Beanstalk because we know pretty well about the environment and what it's doing but people may be running a wide variety of operating systems and doing very different things in their custom scripts.

The patching and custom IAM policies should not be needed anymore once the event based execution work in #354 is finished, I urge you to give that a try and help us get it merged.
